vnext/Microsoft.ReactNative/Views/TextViewManager.cpp

```diff
@@ -55,7 +68,53 @@ class TextShadowNode final : public ShadowNodeBase {
       Super::AddView(*m_firstChildNode, 0);
       m_firstChildNode = nullptr;
     }
+
     Super::AddView(child, index);
+
+    if (auto run = static_cast<ShadowNodeBase &>(child).GetView().try_as<winrt::Run>()) {
+      if (m_isHighlighted) {
+        AddHighlighter(m_ColorValue, run.Text().size());
+      }
+
+      m_prevCursorEnd += run.Text().size();
+    } else if (auto span = static_cast<ShadowNodeBase &>(child).GetView().try_as<winrt::Span>()) {
+      AddNestedTextHighlighter(
+          m_isHighlighted, m_ColorValue, span, static_cast<VirtualTextShadowNode &>(child).m_highlightData.get());
+    }
+  }
+
+  void AddNestedTextHighlighter(
+      bool isParentHighlighted,
+      folly::dynamic parentColor,
+      winrt::Span &span,
+      VirtualTextShadowNode::HighlightData *highData) {
+    if (!highData->isHighlighted && isParentHighlighted) {
+      highData->isHighlighted = true;
+      highData->color = parentColor;
+    }
+
+    for (const auto &el : span.Inlines()) {
+      if (auto run = el.try_as<winrt::Run>()) {
+        if (highData->isHighlighted) {
+          AddHighlighter(highData->color, run.Text().size());
+        }
+
+        m_prevCursorEnd += run.Text().size();
+      } else if (auto spanChild = el.try_as<winrt::Span>()) {
+        AddNestedTextHighlighter(
+            highData->isHighlighted, highData->color, spanChild, highData->data[highData->spanIdx++].get());
+      }
+    }
+  }
+
+  void AddHighlighter(const folly::dynamic &color, size_t runSize) {
+    auto newHigh = winrt::TextHighlighter{};
+    newHigh.Background(SolidColorBrushFrom(color));
+
+    winrt::TextRange newRange{m_prevCursorEnd, static_cast<int32_t>(runSize)};
```

Comment:
this is interesting - the docs say the length of the selection, in Unicode characters - but it doesn't specify if this is supposed to be the number of UTF16 characters, or the number of characters, regardless of encoding. In other words, some Unicode characters will take up more than one 16-bit character (i.e. more than one wchar_t), see "surrogate pairs". Is this number supposed to be the number of UTF16 units, or the number of *characters*? And is the JS property value supposed to be UTF16, or # of chars?

Reply:
I don't know much about this, but here's what I've found:
This - https://www.meziantou.net/how-to-correctly-count-the-number-of-characters-of-a-string.htm - suggests that in .NET a 2-wchar_t character emoji would be of length 2 - I think same goes for UWP. I guess I could test it out and see if a surrogate pair gets highlighted.

At the same time, this - https://github.com/microsoft/Windows-universal-samples/blob/master/Samples/Unicode/cppwinrt/Scenario1_FindId.cpp - shows that processing strings containing surr. pairs involves iterating over its characters calling methods from UnicodeCharacters:: to detect the unusual characters. I've checked and we don't call call any methods from that namespace in our code.. 

So practically, if winrt::TextRange expects the # of logical characters, and if calling complicated_emoji_string.size() will return the number of wchar_t blocks, then I would need to manually check all incoming string to make sure they don't contain the surr. pairs, which is IMO infeasible. At that point I would file a bug (?) with XAML or leave this unsupported in RNW.

Reply:
My main concern is to not expose a buffer overrun vulnerability here. If richedit (which is the thing that backs the xaml text APIs) is expecting number of unicode characters but we give it one more, then richedit could end up reading or writing into invalid memory. I'd be ok with saying "not supported", but not supported should not result in undefined behavior or AVs :)

Reply:
I've tried hardcoding the range length to 1000, and here's what happened:

![image](https://user-images.githubusercontent.com/12816515/96941325-001d6900-1487-11eb-97b1-b506dd832967.png)
vs 
![image](https://user-images.githubusercontent.com/12816515/96941301-eda32f80-1486-11eb-8ae1-59775993d374.png)

You can see that the range just extends until the end of the TextBlock - I didn't see any exception in the logs.

Reply:
Verified under AppVerifier. Resolving.